subscriber/src/daily_operations.rs

```diff
@@ -0,0 +1,227 @@
+    + crate::user_tokens::UserTokensModule
+    + crate::common_storage::CommonStorageModule
+    + energy_query::EnergyQueryModule
+    + multiversx_sc_modules::ongoing_operation::OngoingOperationModule
+{
+    fn perform_service<SC: SubscriberContract<SubSc = Self>>(
+        &self,
+        service_index: usize,
+        user_index: &mut usize,
+        additional_data: ManagedVec<SC::AdditionalDataType>,
+    ) -> OperationCompletionStatus {
+        let mut users_mapper = self.subscribed_users(service_index);
+        let mut total_users = users_mapper.len();
+        let mut progress = self.load_operation::<OperationProgress>();
+        if progress.current_index == 0 {
+            progress.service_index = service_index;
+            progress.current_index = *user_index;
+            progress.additional_data_index = 0;
+        } else {
+            require!(
+                progress.service_index == service_index,
+                "Another operation is in progress"
+            );
+        }
+
+        let fees_contract_address = self.fees_contract_address().get();
+        let energy_threshold = self.energy_threshold().get();
+        let service_info = self.service_info().get().get(service_index);
+        let current_epoch = self.blockchain().get_block_epoch();
+        let additional_data_len = additional_data.len();
+
+        let run_result = self.run_while_it_has_gas(GAS_TO_SAVE_PROGRESS, || {
+            if progress.additional_data_index >= additional_data_len
+                || progress.current_index > total_users
+            {
+                return STOP_OP;
+            }
+
+            let user_data = additional_data.get(*user_index);
+            progress.additional_data_index += 1;
+
+            let user_id = users_mapper.get_by_index(progress.current_index);
+            let opt_user_address = self
+                .user_id()
+                .get_address_at_address(&fees_contract_address, user_id);
+            let user_address = match opt_user_address {
+                Some(address) => address,
+                None => {
+                    users_mapper.swap_remove(&user_id);
+                    total_users -= 1;
+                    return CONTINUE_OP;
+                }
+            };
+
+            progress.current_index += 1;
+
+            let subscription_type = self.subscription_type(user_id, service_index).get();
+            let next_action_epoch =
+                self.get_next_action_epoch(user_id, service_index, subscription_type);
+            if next_action_epoch > current_epoch {
+                return CONTINUE_OP;
+            }
+
+            let subtract_result = self.subtract_payment(
+                &user_address,
+                user_id,
+                &energy_threshold,
+                service_info.payment_type.clone(),
+                fees_contract_address.clone(),
+            );
+            if subtract_result.is_err() {
+                return CONTINUE_OP;
```

Comment:
if this is an error - the whole execution will fail anyway. No need for this check

Reply:
Nope, this is a custom error. If I made it that way, the contract would get stuck if a user didn't deposit tokens for fees.

Reply:
ok :)

Reply:
Nope, it's just an error, but not a VM panic. It simply says "hey, this didn't work".